### PR TITLE
[FEATURE] Backend: Interview Endpoints Implementation

### PR DESCRIPTION
**Pull request by FurkannOzbek**
# Pull Request Title

[FEATURE] Implement Interview Management Endpoints

---

## Description

This PR introduces endpoints for managing interviews, including:

- Creating new interviews
- Retrieving all interviews for a user
- Deleting an interview 
- Updating interview details 

These endpoints ensure that users can manage their interviews associated with job applications while enforcing proper user authorization.

- Problem Solved: Users can now manage their interviews through a set of RESTful endpoints, including the ability to add, update, view, and delete interviews.
- Impact: This change provides users with the functionality needed to handle interviews within the application, with proper authorization checks to ensure that users can only manage their own interviews.
---

## Changes Made

List the main changes in this pull request. Describe what was added, modified, or removed.

- **Frontend:** N/A
- **Backend:** 
  - Implemented routes for GET, POST, PATCH, and DELETE interview operations.
  - Added user authorization checks for each endpoint to ensure that users can only modify or delete interviews related to their applications.
  - Integrated database queries using Knex.js for interacting with the interview and application tables.
  
---

## Testing Steps

Describe how to test this feature or bug fix.

1. **Frontend:** N/A
2. **Backend:**
   - GET interviews: Test by sending a GET request to 'localhost:3001/api/user/applications/:id/interviews' with valid authentication to ensure the correct interviews for the logged-in user are returned.
   - POST interview: Send a POST request with required interview details (type, scheduled_at, etc.) and validate that a new interview is created to 'localhost:3001/api/user/applications/:id/interview'
   - DELETE interview : Send a DELETE request to remove an interview by its ID. Ensure that the request is only allowed for interviews belonging to the authenticated user to  'localhost:3001/api/user/applications/:id/interviews'.
   - PATCH interview Send a PATCH request to update interview details to 'localhost:3001/api/user/applications/:id/interviews'. Verify that only interviews belonging to the logged-in user can be updated.

---

### CODE REVIEW

**Comment by FurkannOzbek** [comment]
I am aware of there are repeating codes about validation for the user_id , If I will have enough time I will create a function for that.

File: api/src/controllers/interviewController.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        .andWhere('application.user_id', user_id)
+
+      if (!interviews || interviews.length === 0) {
+        return res.status(204).json({ message: 'No interviews found' })
+      }
+
+      return res.json(buildInterviewsDto(interviews))
+    } catch (error) {
+      console.error('Database error:', error)
+      return res.status(500).json({ error: 'Database error' })
```
Comment (Natata08) [inline api/src/controllers/interviewController.js:32]: Error without dto is detected 🙂

File: api/src/controllers/interviewController.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        'location',
+        'is_virtual',
+        'created_at',
+        'updated_at',
+      ])
+
+    res.status(200).json(updatedRows)
+  } catch (error) {
+    console.error('error', error)
+    return res.status(500).json({ message: 'internal error' })
```
Comment (Natata08) [inline api/src/controllers/interviewController.js:189]: Here as well

File: api/src/controllers/interviewController.js (hunk 1 above)
Comment (FurkannOzbek) [reply api/src/controllers/interviewController.js:32]: ups :D lemme fix

**Review by Natata08 — DISMISSED** [review]

File: api/src/controllers/interviewController.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    }
+    if (!scheduled_at && !is_virtual) {
+      return res.status(400).json(buildErrorDto('Interview details required'))
+    }
+    try {
+      // Check if the application exists and the user_id matches
+      const application = await knex('application')
+        .where('application_id', applicationId)
+        .select('user_id')
+        .first()
```
Comment (olhasashchuk) [inline api/src/controllers/interviewController.js:56]: I would create the new utils checkApplicationExists, use rules DRY

File: api/src/controllers/interviewController.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  const application_id = parseInt(req.params.id)
+  const { interview_id } = req.body
+  if (!application_id || isNaN(application_id)) {
+    return res.status(400).json(buildErrorDto('Invalid application ID'))
+  }
+  try {
+    const application = await knex('application')
+      .where('application_id', application_id)
+      .select('user_id')
+      .first()
```
Comment (olhasashchuk) [inline api/src/controllers/interviewController.js:108]: also use utils checkApplicationExists here

File: api/src/controllers/interviewController.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    return res.status(400).json(buildErrorDto('Invalid application ID'))
+  }
+  if (!type && !scheduled_at) {
+    return res.status(400).json(buildErrorDto('Interview details required'))
+  }
+  try {
+    const application = await knex('application')
+      .where('application_id', applicationId)
+      .select('user_id')
+      .first()
```
Comment (olhasashchuk) [inline api/src/controllers/interviewController.js:152]: also use utils checkApplicationExists here

File: api/src/controllers/interviewController.js [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+import { buildErrorDto } from '../dtos/errorDto.js'
+
+export const getUserInterviews = async (req, res) => {
+  try {
+    const user_id = req.userInfo.userId
+    const application_id = parseInt(req.params.id)
+    if (!application_id || isNaN(application_id)) {
+      return res.status(400).json(buildErrorDto('Invalid application ID'))
+    }
+    try {
```
Comment (olhasashchuk) [inline api/src/controllers/interviewController.js:13]: how Pankaj recommended, separated for blocks try/catch. You could look an example here: userController.js>patchUserApplicationCompany.js

File: api/src/controllers/interviewController.js [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    if (application.user_id !== user_id) {
+      return res
+        .status(401)
+        .json(
+          buildErrorDto(
+            'Unauthorized access: User does not own this application'
+          )
+        )
+    }
+    const updatedRows = await knex('interview')
```
Comment (olhasashchuk) [inline api/src/controllers/interviewController.js:167]: You should validate the data on the back end. You could use updateField.js

File: api/src/controllers/interviewController.js (hunk 4 above)
Comment (FurkannOzbek) [reply api/src/controllers/interviewController.js:108]: oh yeah I didnt see that one before merge actually did u add it lately? Im gonna use it

**Comment by FurkannOzbek** [comment]
Wrong naming on commit , it includes changes for using utils for checking the application existence with the user_id and seperated try-catch blocks for better error handling

**Review by olhasashchuk — APPROVED** [review]

**Review by Natata08 — APPROVED** [review]

**Pull request merged**